docs/src/tutorial/cython_tutorial.rst

```diff
@@ -132,27 +132,83 @@ them as a Python list.
+of the function (line 22). ::
+
+    if nb_primes > 1000:
+        nb_primes = 1000
+
+As in C, declaring a static array requires knowing the size at compile time.
+We make sure the user doesn't set a value above 1000 (or we'll have a nice
+segmentation fault, just like in C). ::
+
+    len_p = 0  # The number of elements in p
+    n = 2
+    while len_p < nb_primes:
 
 Lines 7-9 set up for a loop which will test candidate numbers for primeness
-until the required number of primes has been found. Lines 11-12, which try
-dividing a candidate by all the primes found so far, are of particular
-interest. Because no Python objects are referred to, the loop is translated
-entirely into C code, and thus runs very fast.
-
-When a prime is found, lines 14-15 add it to the p array for fast access by
-the testing loop, and line 16 adds it to the result list. Again, you'll notice
-that line 16 looks very much like a Python statement, and in fact it is, with
-the twist that the C parameter ``n`` is automatically converted to a Python
-object before being passed to the append method. Finally, at line 18, a normal
+until the required number of primes has been found. ::
+
+    # Is n prime?
+    for i in p[:len_p]:
+        if n % i == 0:
+            break
+
+Lines 11-12, which try dividing a candidate by all the primes found so far,
+are of particular interest. Because no Python objects are referred to,
+the loop is translated entirely into C code, and thus runs very fast.
+You will notice the way we iterate over the ``p`` C array.  ::
+
+    for i in p[:len_p]:
+
+The loop gets translated into C code transparently. No more ugly C for loops!
+Well don't forget how to loop in C style with integers yet, you might need it someday.
+If you don't use ``:len_p`` then Cython will loop over the 1000 elements of
+the array (it won't go out of bounds and give a segmentation fault). ::
+
+    # If no break occurred in the loop
+    else:
+        p[len_p] = n
+        len_p += 1
+    n += 1
+
+If no breaks occurred, it means that we found a prime, and the block of code
+after the ``else`` line 16 will be executed. We add the prime found to ``p``.
+If you find having a else after a for loop strange, just know that it's a
+hidden secret of the python syntax, and actually doesn't exist in C!
+But since Cython is made to be written with the Python syntax, it'll
+work out, as if you wrote Python code, but at C speed in this case.
+If the for...else syntax still confuses you, see this excellent
+`blog post <https://shahriar.svbtle.com/pythons-else-clause-in-loops>`_. ::
+
+    # Let's put the result in a python list:
+    result_as_list  = [prime for prime in p[:len_p]]
+    return result_as_list
+
+Line 22, before returning the result, we need to convert our C array into a
+Python list, because Python can't read C arrays. Note that Cython handle
+for you the conversion of quite some types between C and Python (you can
```

Comment:
"Note that ... " -> "Cython can automatically convert many C types from and to Python types, as described in the documentation on [type conversion]."